R/seq_linter.R

```diff
@@ -1,19 +1,25 @@
 #' Sequence linter
 #'
-#' Check for `1:length(...)`, `1:nrow(...)`, `1:ncol(...)`, `1:NROW(...)` and `1:NCOL(...)` expressions.
+#' Check for `1:length(...)`, `1:nrow(...)`, `1:ncol(...)`, `1:NROW(...)` and `1:NCOL(...)` expressions in base-R.
+#' Additionally, `1:n()` (from dplyr package) and `1:.N` (from data.table package) are also checked.
 #' These often cause bugs when the right-hand side is zero.
 #' It is safer to use [base::seq_len()] or [base::seq_along()] instead.
 #'
 #' @evalRd rd_tags("seq_linter")
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
 seq_linter <- function() {
-  bad_funcs <- c("length", "nrow", "ncol", "NROW", "NCOL", "dim")
+  bad_funcs <- c("length", "n", "nrow", "ncol", "NROW", "NCOL", "dim")
 
+  # `.N` from {data.table} is special since it's not a function but a symbol
   xpath <- glue::glue("//expr[
     expr[NUM_CONST[text() =  '1' or text() =  '1L']]
     and OP-COLON
-    and expr[expr[(expr|self::*)[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(bad_funcs)} ]]]]
+    and (
+          expr[expr[(expr|self::*)[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(bad_funcs)} ]]]]
+          or
+          expr[SYMBOL='.N']
```

Comment:
oh, I didn't know this works!

usually we'd do

SYMBOL[text() = '.N']

let me check the docs to ensure there's nothing subtle about this XPath usage

Reply:
nice! looks like this is legit:

> If one object to be compared is a node-set and the other is a string, then the comparison will be true if and only if there is a node in the node-set such that the result of performing the comparison on the string-value of the node and the other string is true.

https://www.w3.org/TR/1999/REC-xpath-19991116/#section-Expressions

Its for != where things are a bit trickier:

https://stackoverflow.com/a/4629660/3576984